### PR TITLE
Improve comparison by using internal typescript methods

### LINKED ISSUE #380: Compatibility check finds something that doesn’t exist in my project
Ref: https://community.grafana.com/t/compatibility-check-finds-something-that-doesnt-exist-in-my-project/111182

---

Hi all, I'm getting an [error](https://github.com/nicfv/Psychart/actions/runs/7398043595/job/20126338962#step:6:22) in my compatibility check workflow that is not part of my project:

```
1) Changed `PanelPlugin.setPanelOptions` used in `src/module.ts:267`
Parameter type changed:

+ Added - Removed

-declare type PanelOptionsSupplier<TOptions> = (builder: PanelOptionsEditorBuilder<TOptions>, context: StandardEditorContext<TOptions>) => void;
+type PanelOptionsSupplier<TOptions> = (builder: PanelOptionsEditorBuilder<TOptions>, context: StandardEditorContext<TOptions>) => void;
```

First of all, my [`module.ts`](https://github.com/nicfv/Psychart/blob/ce5c93299fac92de286d3db728d2bf62648033cb/src/module.ts#L1) file is only 266 lines, and I also searched for "PanelOptionsSupplier" and it is not anywhere in my file. I also found a GitHub [issue](https://github.com/grafana/levitate/issues/231) that may be loosely related. Other than that, I'm stuck. Can someone help me track this down please?

### PR DESCRIPTION
**Pull request by academo**
**What this PR does / why we need it**:

This PR is similar to https://github.com/grafana/levitate/pull/361 but it doesn't put all the trust in the typescript internal checker.

In the previous PR, the internal ts checker would report errors between two identical packages because of internal ts nuances that created fake incompatible types.

The newer approach is to trust the typescript internal methods if it says two types are compatible but, if it says they are not, we always fallback to text comparison.

Additionally I added a test case to track an edge case with typescript autogenerated types suffixes.

**Which issue(s) this PR fixes**:

It closes https://github.com/grafana/levitate/issues/380

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: src/commands/compare/classes.test.ts [hunk 1]
```diff
@@ -367,7 +367,7 @@ describe('Compare classes', () => {
     expect(Object.keys(comparison.removals).length).toBe(0);
   });
 
-  test('NEW ARGUMENT FOR A CLASS METHOD - adding a new positional argument to any of the methods should trigger a breaking change', () => {
+  test('NEW ARGUMENT FOR A CLASS METHOD - adding a new positional argument to any of the methods should not trigger a breaking change', () => {
```
Comment (academo) [inline src/commands/compare/classes.test.ts:370]: Why?

This used to be a breaking change for the `Foo` type but this was because of the text comparison. For typescript, calling  a function and missing positional parameters is not a problem thus adding a new one is not a breaking change.

File: src/commands/compare/classes.test.ts [hunk 2]
```diff
@@ -504,7 +504,7 @@ describe('Compare classes', () => {
     `;
     const comparison = testCompare(prev, current);
 
-    expect(Object.keys(comparison.changes).length).toBe(2); // Foo and Foo<T> changed
+    expect(Object.keys(comparison.changes)).toEqual(['Foo.T']);
```
Comment (academo) [inline src/commands/compare/classes.test.ts:507]: This used to be a breaking change for `Foo` but in reality only the generic type `Foo.T` changed.

File: src/commands/compare/functions.test.ts [hunk 3]
```diff
@@ -105,7 +105,7 @@ describe('Compare functions', () => {
     `;
     const comparison = testCompare(prev, current);
 
-    expect(Object.keys(comparison.changes)).toEqual(['foo']);
+    expect(Object.keys(comparison.changes).length).toBe(0);
```
Comment (academo) [inline src/commands/compare/functions.test.ts:108]: Same as with the class method. New positional arguments are not breaking changes.

File: fixtures/compare/props-suffix/current.d.ts
Comment (academo) [inline fixtures/compare/props-suffix/current.d.ts:1]: big fixture file extracted from a grafana/ui PR to test an edge case

File: fixtures/compare/props-suffix/previous.d.ts
Comment (academo) [inline fixtures/compare/props-suffix/previous.d.ts:1]: big fixture file extracted from a grafana/ui PR to test an edge case

File: src/commands/compare/compare.ts [hunk 4]
```diff
@@ -140,6 +140,22 @@ export function getFunctionParametersDiff({
...
+  // which is better than text comparison
+  // but it is not 100% reliable so we fallback to text comparison
+  const checker = prev.program.getTypeChecker();
+  const isComparable = checker.isTypeAssignableTo(
+    checker.getTypeAtLocation(prevDeclaration),
+    checker.getTypeAtLocation(currentDeclaration)
+  );
+
+  // we can trust if the typescript compiler calls them comparable
+  // but we can't trust when they are not
```
Comment (leventebalogh) [inline src/commands/compare/compare.ts:154]: **Nit:** I think it would be helpful for the future to give an example for when TS compiler gives a false positive

File: src/commands/compare/classes.test.ts (hunk 1 above)
Comment (leventebalogh) [reply src/commands/compare/classes.test.ts:370]: Hmm, are we sure about this? If we add a new positional parameter to a method (which is not optional), then probably it means that the method is relying on it, and old implementations not providing that parameter will break, no? Am I missing something?

Screenshot 2024-03-19 at 08 30 58

File: src/commands/compare/classes.test.ts [hunk 5]
```diff
@@ -432,16 +432,16 @@ describe('Compare classes', () => {
...
 
         // NO CHANGE
+        methodA(colName: number): number; // the name of the argument changed but it's the same type
         methodC(col: string): string;
       }
     `;
     const comparison = testCompare(prev, current);
 
-    expect(Object.keys(comparison.changes).length).toBe(3); // Foo, Foo.methodA and Foo.methodB changed
+    expect(Object.keys(comparison.changes)).toEqual(['Foo', 'Foo.methodB']);
```
[the remark this answers is not in the document]
Comment (leventebalogh) [reply src/commands/compare/classes.test.ts:444]: Nice 👏

File: src/commands/compare/classes.test.ts [hunk 6]
```diff
@@ -485,9 +485,9 @@ describe('Compare classes', () => {
     expect(Object.keys(comparison.removals).length).toBe(1); // Foo.methodA removed
   });
 
- test('changing the generic should trigger a change', () => {
+  test('changing the generic should trigger a change in the generic type', () => {
     const prev = `
-      export declare class Foo<T = any> {
+      export declare class Foo<T = any > {
```
Comment (leventebalogh) [inline src/commands/compare/classes.test.ts:490]: **Nit:** intended whitespace?

File: src/commands/compare/compare.test.ts [hunk 7]
```diff
@@ -27,4 +31,21 @@ describe('Utils compare tests', () => {
...
+    const fixturePath = path.resolve(__dirname, '../../../fixtures/compare/props-suffix/');
+    const current = path.resolve(fixturePath, 'current.d.ts');
+    const prev = path.resolve(fixturePath, 'previous.d.ts');
+    const ignoredExports = {};
+    const comparison = compareExports(prev, current, ignoredExports);
+
+    // this is a false positive
+    // there shouldn't be any change between the fixtures
+    // keeping it here to document it
+    expect(Object.keys(comparison.changes)).toEqual(['VizRepeater.componentDidUpdate']);
```
Comment (leventebalogh) [inline src/commands/compare/compare.test.ts:45]: Is this something that we ideally would like to fix in the future?

File: src/commands/compare/functions.test.ts [hunk 8]
```diff
@@ -123,4 +123,50 @@ describe('Compare functions', () => {
...
+      export function foo(newNameForBar: number) {
+        bar;
+      }
+    `;
+    const comparison = testCompare(prev, current);
+
+    expect(Object.keys(comparison.changes).length).toBe(0);
+    expect(Object.keys(comparison.additions).length).toBe(0);
+    expect(Object.keys(comparison.removals).length).toBe(0);
+  });
```
Comment (leventebalogh) [inline src/commands/compare/functions.test.ts:171]: Nice ones 👏

**Review by leventebalogh — COMMENTED** [review]
Great enhancements 🙌  
Left one question regarding the positional arguments.

File: src/commands/compare/compare.test.ts (hunk 7 above)
Comment (academo) [reply src/commands/compare/compare.test.ts:45]: Yes. I am leaving this test case for us to remember in the future that this is a problem

**Review by leventebalogh — APPROVED** [review]
🚢🇮🇹
